File: app/src/main/java/wangdaye/com/geometricweather/theme/resource/providers/ChronusResourceProvider.java
package wangdaye.com.geometricweather.theme.resource.providers;

import android.animation.Animator;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.Icon;
import android.net.Uri;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.core.content.res.ResourcesCompat;

import java.util.ArrayList;
import java.util.List;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.common.basic.models.weather.WeatherCode;
import wangdaye.com.geometricweather.theme.resource.utils.Constants;
import wangdaye.com.geometricweather.theme.resource.utils.ResourceUtils;

public class ChronusResourceProvider extends ResourceProvider {

    private final ResourceProvider mDefaultProvider;

    private Context mContext;
    private String mProviderName;
    @Nullable private Drawable mIconDrawable;

    public ChronusResourceProvider(@NonNull Context c, @NonNull String pkgName,
                                   @NonNull ResourceProvider defaultProvider) {
        mDefaultProvider = defaultProvider;

        try {
            mContext = c.createPackageContext(
                    pkgName, Context.CONTEXT_INCLUDE_CODE | Context.CONTEXT_IGNORE_SECURITY);

            PackageManager manager = mContext.getPackageManager();
            ApplicationInfo info = manager.getApplicationInfo(pkgName, PackageManager.GET_META_DATA);
            mProviderName = manager.getApplicationLabel(info).toString();

            mIconDrawable = mContext.getApplicationInfo().loadIcon(mContext.getPackageManager());
        } catch (Exception e) {
            buildDefaultInstance(c);
        }
    }

    private void buildDefaultInstance(@NonNull Context c) {
        mContext = c.getApplicationContext();
        mProviderName = c.getString(R.string.geometric_weather);
        mIconDrawable = mDefaultProvider.getProviderIcon();
    }

    @NonNull
    public static List<ChronusResourceProvider> getProviderList(@NonNull Context context,
                                                                @NonNull ResourceProvider defaultProvider) {
        List<ChronusResourceProvider> providerList = new ArrayList<>();

        List<ResolveInfo> infoList = context.getPackageManager().queryIntentActivities(
                new Intent(Intent.ACTION_MAIN).addCategory(Constants.CATEGORY_CHRONUS_ICON_PACK),
                PackageManager.GET_RESOLVED_FILTER
        );
        for (ResolveInfo info : infoList) {
            providerList.add(
                    new ChronusResourceProvider(
                            context,
                            info.activityInfo.applicationInfo.packageName,
                            defaultProvider
                    )
            );
        }

        return providerList;
    }

    public static boolean isChronusIconProvider(@NonNull Context context, @NonNull String packageName) {
        List<ResolveInfo> infoList = context.getPackageManager().queryIntentActivities(
                new Intent(Intent.ACTION_MAIN).addCategory(Constants.CATEGORY_CHRONUS_ICON_PACK),
                PackageManager.GET_RESOLVED_FILTER
        );
        for (ResolveInfo info : infoList) {
            if (packageName.equals(info.activityInfo.applicationInfo.packageName)) {
                return true;
            }
        }

        return false;
    }

    @Override
    public String getPackageName() {
        return mContext.getPackageName();
    }

    @Override
    public String getProviderName() {
        return mProviderName;
    }

    @Override
    public Drawable getProviderIcon() {
        if (mIconDrawable == null) {
            return getWeatherIcon(WeatherCode.CLEAR, true);
        } else {
            return mIconDrawable;
        }
    }

    // weather icon.

    @NonNull
    @Override
    public Drawable getWeatherIcon(WeatherCode code, boolean dayTime) {
        try {
            return ResourceUtils.nonNull(
                    getDrawable(getWeatherIconName(code, dayTime))
            );
        } catch (Exception ignore) {

        }

        return mDefaultProvider.getWeatherIcon(code, dayTime);
    }

    @NonNull
    @Override
    public Uri getWeatherIconUri(WeatherCode code, boolean dayTime) {
        String resName = getWeatherIconName(code, dayTime);
        int resId = getResId(mContext, resName, "drawable");
        if (resId != 0) {
            return getDrawableUri(resName);
        } else {
            return mDefaultProvider.getWeatherIconUri(code, dayTime);
        }
    }

    @Override
    public Drawable[] getWeatherIcons(WeatherCode code, boolean dayTime) {
        return new Drawable[] {getWeatherIcon(code, dayTime), null, null};
    }

    @Nullable
    private Drawable getDrawable(@NonNull String resName) {
        try {
            return ResourcesCompat.getDrawable(
                    mContext.getResources(),
                    ResourceUtils.nonNull(getResId(mContext, resName, "drawable")),
                    null
            );
        } catch (Exception e) {
            return null;
        }
    }

    String getWeatherIconName(WeatherCode code, boolean daytime) {
        return "weather" + Constants.SEPARATOR + getWeatherIconCode(code, daytime);
    }

    private String getWeatherIconCode(WeatherCode code, boolean daytime) {
        switch (code) {
            case CLEAR:
                return daytime ? "32" : "31";

            case PARTLY_CLOUDY:
                return daytime ? "30" : "29";

            case CLOUDY:
                return "26";

            case RAIN:
                return "11";

            case SNOW:
                return "16";

            case WIND:
                return "24";

            case FOG:
                return "20";

            case HAZE:
                return "19";

            case SLEET:
                return "5";

            case HAIL:
                return "17";

            case THUNDER:
                return daytime ? "37" : "47";

            case THUNDERSTORM:
                return daytime ? "38" : "45";
        }
        return "na";
    }

    // animator.

    @Override
    public Animator[] getWeatherAnimators(WeatherCode code, boolean dayTime) {
        return new Animator[] {null, null, null};
    }

    // minimal icon.

    @NonNull
    @Override
    public Drawable getMinimalLightIcon(WeatherCode code, boolean dayTime) {
        return mDefaultProvider.getMinimalLightIcon(code, dayTime);
    }

    @NonNull
    @Override
    public Uri getMinimalLightIconUri(WeatherCode code, boolean dayTime) {
        return mDefaultProvider.getMinimalLightIconUri(code, dayTime);
    }

    @NonNull
    @Override
    public Drawable getMinimalGreyIcon(WeatherCode code, boolean dayTime) {
        return mDefaultProvider.getMinimalGreyIcon(code, dayTime);
    }

    @NonNull
    @Override
    public Uri getMinimalGreyIconUri(WeatherCode code, boolean dayTime) {
        return mDefaultProvider.getMinimalGreyIconUri(code, dayTime);
    }

    @NonNull
    @Override
    public Drawable getMinimalDarkIcon(WeatherCode code, boolean dayTime) {
        return mDefaultProvider.getMinimalDarkIcon(code, dayTime);
    }

    @NonNull
    @Override
    public Uri getMinimalDarkIconUri(WeatherCode code, boolean dayTime) {
        return mDefaultProvider.getMinimalDarkIconUri(code, dayTime);
    }

    @NonNull
    @Override
    public Drawable getMinimalXmlIcon(WeatherCode code, boolean dayTime) {
        return mDefaultProvider.getMinimalXmlIcon(code, dayTime);
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    @NonNull
    @Override
    public Icon getMinimalIcon(WeatherCode code, boolean dayTime) {
        return mDefaultProvider.getMinimalIcon(code, dayTime);
    }

    // shortcut.

    @NonNull
    @Override
    public Drawable getShortcutsIcon(WeatherCode code, boolean dayTime) {
        return mDefaultProvider.getShortcutsIcon(code, dayTime);
    }

    @NonNull
    @Override
    public Drawable getShortcutsForegroundIcon(WeatherCode code, boolean dayTime) {
        return mDefaultProvider.getShortcutsForegroundIcon(code, dayTime);
    }

    @NonNull
    @Override
    public Drawable getSunDrawable() {
        return getWeatherIcon(WeatherCode.CLEAR, true);
    }

    @NonNull
    @Override
    public Drawable getMoonDrawable() {
        return getWeatherIcon(WeatherCode.CLEAR, false);
    }
}

File: app/src/main/kotlin/org/breezyweather/ui/theme/resource/providers/ChronusResourceProvider.kt
/*
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.ui.theme.resource.providers

import android.animation.Animator
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.drawable.Drawable
import android.graphics.drawable.Icon
import android.net.Uri
import android.os.Build
import androidx.annotation.RequiresApi
import androidx.core.content.res.ResourcesCompat
import breezyweather.domain.weather.reference.WeatherCode
import org.breezyweather.R
import org.breezyweather.ui.theme.resource.utils.Constants

class ChronusResourceProvider(
    c: Context,
    pkgName: String,
    private val mDefaultProvider: ResourceProvider,
) : ResourceProvider() {
    private lateinit var mContext: Context
    override var providerName: String? = null
    private var mIconDrawable: Drawable? = null

    init {
        try {
            mContext = c.createPackageContext(
                pkgName,
                Context.CONTEXT_INCLUDE_CODE or Context.CONTEXT_IGNORE_SECURITY
            )
            val manager = mContext.packageManager
            val info = manager.getApplicationInfo(pkgName, PackageManager.GET_META_DATA)
            providerName = manager.getApplicationLabel(info).toString()
            mIconDrawable = mContext.applicationInfo.loadIcon(mContext.packageManager)
        } catch (e: Exception) {
            buildDefaultInstance(c)
        }
    }

    private fun buildDefaultInstance(c: Context) {
        mContext = c.applicationContext
        providerName = c.getString(R.string.breezy_weather)
        mIconDrawable = mDefaultProvider.providerIcon
    }

    override val packageName: String
        get() = mContext.packageName
    override val providerIcon: Drawable
        get() = mIconDrawable ?: getWeatherIcon(WeatherCode.CLEAR, true)

    // weather icon.
    override fun getWeatherIcon(code: WeatherCode?, dayTime: Boolean): Drawable {
        try {
            return getDrawable(getWeatherIconName(code, dayTime))!!
        } catch (ignore: Exception) { }
        return mDefaultProvider.getWeatherIcon(code, dayTime)
    }

    override fun getWeatherIconUri(code: WeatherCode?, dayTime: Boolean): Uri {
        val resName = getWeatherIconName(code, dayTime)
        val resId = getResId(mContext, resName, "drawable")
        return if (resId != 0) {
            getDrawableUri(resName)
        } else {
            mDefaultProvider.getWeatherIconUri(code, dayTime)
        }
    }

    override fun getWeatherIcons(code: WeatherCode?, dayTime: Boolean): Array<Drawable?> {
        return arrayOf(getWeatherIcon(code, dayTime), null, null)
    }

    private fun getDrawable(resName: String): Drawable? {
        return try {
            ResourcesCompat.getDrawable(
                mContext.resources,
                getResId(mContext, resName, "drawable"),
                null
            )
        } catch (e: Exception) {
            null
        }
    }

    fun getWeatherIconName(code: WeatherCode?, daytime: Boolean): String {
        return "weather" + Constants.SEPARATOR + getWeatherIconCode(code, daytime)
    }

    private fun getWeatherIconCode(code: WeatherCode?, daytime: Boolean): String {
        return when (code) {
            WeatherCode.CLEAR -> if (daytime) "32" else "31"
            WeatherCode.PARTLY_CLOUDY -> if (daytime) "30" else "29"
            WeatherCode.CLOUDY -> "26"
            WeatherCode.RAIN -> "12"
            WeatherCode.SNOW -> "16"
            WeatherCode.WIND -> "24"
            WeatherCode.FOG -> "20"
            WeatherCode.HAZE -> "21"
            WeatherCode.SLEET -> "5"
            WeatherCode.HAIL -> "17"
            WeatherCode.THUNDER -> "4"
            WeatherCode.THUNDERSTORM -> "4"
            else -> "na"
        }
    }

    // animator.
    override fun getWeatherAnimators(code: WeatherCode?, dayTime: Boolean): Array<Animator?> {
        return arrayOf(null, null, null)
    }

    // minimal icon.
    override fun getMinimalLightIcon(code: WeatherCode?, dayTime: Boolean): Drawable {
        return mDefaultProvider.getMinimalLightIcon(code, dayTime)
    }

    override fun getMinimalLightIconUri(code: WeatherCode?, dayTime: Boolean): Uri {
        return mDefaultProvider.getMinimalLightIconUri(code, dayTime)
    }

    override fun getMinimalGreyIcon(code: WeatherCode?, dayTime: Boolean): Drawable {
        return mDefaultProvider.getMinimalGreyIcon(code, dayTime)
    }

    override fun getMinimalGreyIconUri(code: WeatherCode?, dayTime: Boolean): Uri {
        return mDefaultProvider.getMinimalGreyIconUri(code, dayTime)
    }

    override fun getMinimalDarkIcon(code: WeatherCode?, dayTime: Boolean): Drawable {
        return mDefaultProvider.getMinimalDarkIcon(code, dayTime)
    }

    override fun getMinimalDarkIconUri(code: WeatherCode?, dayTime: Boolean): Uri {
        return mDefaultProvider.getMinimalDarkIconUri(code, dayTime)
    }

    override fun getMinimalXmlIcon(code: WeatherCode?, dayTime: Boolean): Drawable {
        return mDefaultProvider.getMinimalXmlIcon(code, dayTime)
    }

    override fun getMinimalIcon(code: WeatherCode?, dayTime: Boolean): Icon {
        return mDefaultProvider.getMinimalIcon(code, dayTime)
    }

    // shortcut.
    override fun getShortcutsIcon(code: WeatherCode?, dayTime: Boolean): Drawable {
        return mDefaultProvider.getShortcutsIcon(code, dayTime)
    }

    override fun getShortcutsForegroundIcon(code: WeatherCode?, dayTime: Boolean): Drawable {
        return mDefaultProvider.getShortcutsForegroundIcon(code, dayTime)
    }

    override val sunDrawable: Drawable
        get() = getWeatherIcon(WeatherCode.CLEAR, true)
    override val moonDrawable: Drawable
        get() = getWeatherIcon(WeatherCode.CLEAR, false)

    companion object {
        fun getProviderList(
            context: Context,
            defaultProvider: ResourceProvider,
        ): List<ChronusResourceProvider> {
            val providerList = mutableListOf<ChronusResourceProvider>()
            val infoList = context.packageManager.queryIntentActivities(
                Intent(Intent.ACTION_MAIN).addCategory(Constants.CATEGORY_CHRONUS_ICON_PACK),
                PackageManager.GET_RESOLVED_FILTER
            )
            for (info in infoList) {
                providerList.add(
                    ChronusResourceProvider(
                        context,
                        info.activityInfo.applicationInfo.packageName,
                        defaultProvider
                    )
                )
            }
            return providerList
        }

        fun isChronusIconProvider(context: Context, packageName: String): Boolean {
            val infoList = context.packageManager.queryIntentActivities(
                Intent(Intent.ACTION_MAIN).addCategory(Constants.CATEGORY_CHRONUS_ICON_PACK),
                PackageManager.GET_RESOLVED_FILTER
            )
            for (info in infoList) {
                if (packageName == info.activityInfo.applicationInfo.packageName) {
                    return true
                }
            }
            return false
        }
    }
}
